test/e2e/utils/run_cov_binary.sh

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+# Copyright 2024 Antrea Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is meant to wrap a coverage instrumented binary.
+# When the main process is killed, the script will start sleeping until a signal
+# is received, providing the opportunity for the user to retrieve code coverage
+# files.
+
+set -x
+
+"$@"
+
+while true; do
+    # Sleep in the background, to allow exiting as soon as a signal is received
+    sleep 5 & wait $!
+done
```

Comment:
I think I figured out the root cause for https://github.com/antrea-io/antrea/issues/6318

This code is not handling the SIGTERM signal correctly, which means that the antrea-agent Pod takes a full 30s to terminate when it is deleted. This means that when we delete the antrea-agent Pod in a test case, it has the potential of impacting subsequent test cases, depending on how these test cases are written.

This is my bad since I came up with the code for the script. Can you replace it with the following code? It should solve the issue.

```bash
set -euo pipefail

TEST_PID=

function quit {
    echo "Received signal, exiting"
    # While not strictly required, it is better to try to terminate the test process gracefully if it is still running.
    if [ "$TEST_PID" != "" ]; then
        echo "Sending signal to test process"
        kill -SIGTERM $TEST_PID > /dev/null 2>&1 || true
        wait $TEST_PID
        echo "Test process exited gracefully"
    fi
    exit 0
}

# This is necessary: without an explicit signal handler, the SIGTERM signal will be discarded when running
# as a process with PID 1 in a container.
trap 'quit' SIGTERM

# Run the test process in the background, to allow exiting when a signal is received.
"$@" &
TEST_PID=$!
wait $TEST_PID
TEST_PID=

# Sleep in the background, to allow exiting as soon as a signal is received.
sleep infinity & wait
```

Reply:
Looks like this way will impact the coverage collection?
```
 2024/05/11 15:31:41 Error when gracefully exit antrea agent: error when gracefully exiting Antrea Agent: error 
when copying antctl coverage files: error when running this find command '[bash -c find /tmp/coverage  -mindepth 
1]' on Pod 'antrea-agent-cp7c4', stderr: <>, err: <Internal error occurred: error executing command in container: 
failed to exec in container: failed to start exec 
"1d044df8621e3a24943111287da3529bb53e25c04a2e8c8da3eecb1625916618": OCI runtime exec failed: exec 
failed: cannot exec in a stopped container: unknown>
```

Reply:
@luolanzone Where are these logs from?

Reply:
I have also seen these errors in the logs for one of the Github CI jobs:

```
2024-05-11T15:06:39.0693719Z I0511 15:06:39.068915   23828 framework.go:2651] Sending SIGINT to 'antrea-agent-coverage'
2024-05-11T15:06:39.1196604Z I0511 15:06:39.119191   23828 framework.go:2657] Copying coverage files from Pod 'antrea-agent-j92js'
2024-05-11T15:06:42.3927758Z I0511 15:06:42.391304   23828 framework.go:2797] Copying file: covmeta.ede75673d2ae4194072b9dd9af8339de
2024-05-11T15:06:42.4480081Z I0511 15:06:42.447436   23828 framework.go:2737] Coverage file not available yet for copy: cannot retrieve content of file '/tmp/coverage/covmeta.ede75673d2ae4194072b9dd9af8339de' from Pod 'antrea-agent-j92js', stderr: <>, err: <Internal error occurred: error executing command in container: failed to exec in container: failed to create exec "55149056a7d640b206c4e2daf48e4c25985b2df698c89c225e0731372cd38449": task a4707bcd87f8f641de784cabd92eeec4adf3932f00874264ae59fd138ab4dc67 not found: not found>
2024-05-11T15:06:43.3923117Z I0511 15:06:43.391975   23828 framework.go:2797] Copying file: covmeta.ede75673d2ae4194072b9dd9af8339de
```

```
2024-05-11T15:31:41.3860016Z I0511 15:31:41.385498   23828 framework.go:2657] Copying coverage files from Pod 'antrea-agent-cp7c4'
2024-05-11T15:31:41.4442434Z 2024/05/11 15:31:41 Error when gracefully exit antrea agent: error when gracefully exiting Antrea Agent: error when copying antctl coverage files: error when running this find command '[bash -c find /tmp/coverage  -mindepth 1]' on Pod 'antrea-agent-cp7c4', stderr: <>, err: <Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "1d044df8621e3a24943111287da3529bb53e25c04a2e8c8da3eecb1625916618": OCI runtime exec failed: exec failed: cannot exec in a stopped container: unknown>
```

This is the job: https://github.com/antrea-io/antrea/actions/runs/9044365104/job/24853065716?pr=6090

These errors are confusing as they seem to indicate that the container doesn't exist anymore. Which should not be the case since at that point we have killed the process, but not the Pod / container. The first one in particular is confusing for me, because it seems that we succeed in copying the file on the second attempt. Any idea @shikharish ?

Reply:
We can run the test again after @shikharish disables the `TestConnectivity/testOVSRestartSameNode` test case, but it's making me a bit uneasy that we are even seeing these errors now.

Reply:
I found that the failure is in `TestWireGuard` and only due to the flag `--feature-gates AllAlpha=true`.

Reply:
This is indeed the only case in which we see it happen in CI, but I don't think it has anything to do with `TestWireGuard`. `TestWireGuard` is actually skipped when all features are enabled, as it is not compatible with Multicast. The errors happen when we collect coverage data at the end of the test suite: https://github.com/antrea-io/antrea/blob/6a7c60610c3baedbc427b786edc11eea6b4a5b06/test/e2e/fixtures.go#L750

We need to figure out what's going on, and it seems to happen consistently in CI. I wasn't able to reproduce locally by running the following command: `go test -timeout=10m -v -count=1 -run=TestWireGuard  antrea.io/antrea/test/e2e -provider=kind -coverage=true -deploy-antrea=false`.

This means that we will need to find a way to extract more information from the CI job, in order to troubleshoot.
It looks to me that we are always able to 1) retrieve the PID for the Agent process, and 2) kill the process. However, something is not working when we retrieve coverage data: one of the container seems to have been stopped. If you can find a way to do a `kubectl describe` on that specific Pod (refer to `waitForAntreaDaemonSetPods`), this may provide us with enough information. I think the right place to do it is in `killProcessAndCollectCovFiles`. If `collectCovFiles` returns an error, run `kubectl describe` for that Pod, and print the contents of stdout.
Please make these changes in a separate commit; these changes may not need to be temporary. If done well we could keep them for future troubleshooting.

Reply:
@antoninbas I reproduced the error locally, here is the `kubectl describe` output on the Pod. 
```
I0515 03:17:30.996186 1288821 framework.go:2661] 
Name:                 antrea-controller-697f964f85-n8czk
Namespace:            kube-system
Priority:             2000000000
Priority Class Name:  system-cluster-critical
Service Account:      antrea-controller
Node:                 kind-worker/172.18.0.3
Start Time:           Tue, 14 May 2024 21:42:10 +0000
Labels:               app=antrea
                      component=antrea-controller
                      pod-template-hash=697f964f85
Annotations:          checksum/config: cd88d185adb537e02580d3120f377ea3ac31c96a80b9ff6a074fd89d9102c632
Status:               Running
IP:                   172.18.0.3
IPs:
  IP:           172.18.0.3
Controlled By:  ReplicaSet/antrea-controller-697f964f85
Containers:
  antrea-controller:
    Container ID:  containerd://0f4a726ec57f3e4fbd4b5a5d8de5a97fd73338edc55ef0ab4041ebc0bc7b1124
    Image:         antrea/antrea-controller-ubuntu-coverage:latest
    Image ID:      docker.io/library/import-2024-05-14@sha256:e7f52024e3c096a1412e4aa012b80e85808a37b974e2c90765d53abdff5dee79
    Port:          10349/TCP
    Host Port:     10349/TCP
    Args:
      antrea-controller-coverage
      --config=/etc/antrea/antrea-controller.conf
      --logtostderr=false
      --log_dir=/var/log/antrea
      --alsologtostderr
      --log_file_max_size=100
      --log_file_max_num=4
      --v=4
    State:          Running
      Started:      Tue, 14 May 2024 21:47:11 +0000
    Last State:     Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 14 May 2024 21:46:21 +0000
      Finished:     Tue, 14 May 2024 21:47:11 +0000
    Ready:          True
    Restart Count:  3
    Requests:
      cpu:      200m
      memory:   100Mi
    Liveness:   http-get https://localhost:api/livez delay=0s timeout=5s period=10s #success=1 #failure=5
    Readiness:  http-get https://localhost:api/readyz delay=5s timeout=5s period=10s #success=1 #failure=5
    Environment:
      POD_NAME:                antrea-controller-697f964f85-n8czk (v1:metadata.name)
      POD_NAMESPACE:           kube-system (v1:metadata.namespace)
      NODE_NAME:                (v1:spec.nodeName)
      SERVICEACCOUNT_NAME:      (v1:spec.serviceAccountName)
      ANTREA_CONFIG_MAP_NAME:  antrea-config
    Mounts:
      /etc/antrea/antrea-controller.conf from antrea-config (ro,path="antrea-controller.conf")
      /var/log/antrea from host-var-log-antrea (rw)
      /var/run/antrea/antrea-controller-tls from antrea-controller-tls (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-9mpbf (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       True 
  ContainersReady             True 
  PodScheduled                True 
Volumes:
  antrea-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      antrea-config
    Optional:  false
  antrea-controller-tls:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  antrea-controller-tls
    Optional:    true
  host-var-log-antrea:
    Type:          HostPath (bare host directory volume)
    Path:          /var/log/antrea
    HostPathType:  DirectoryOrCreate
  kube-api-access-9mpbf:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 CriticalAddonsOnly op=Exists
                             node-role.kubernetes.io/control-plane:NoSchedule
                             node-role.kubernetes.io/master:NoSchedule
                             node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 0s
Events:
  Type     Reason            Age                    From               Message
  ----     ------            ----                   ----               -------
  Warning  FailedScheduling  5m21s                  default-scheduler  0/3 nodes are available: 3 node(s) had untolerated taint {node.kubernetes.io/not-ready: }. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
  Normal   Scheduled         5m20s                  default-scheduler  Successfully assigned kube-system/antrea-controller-697f964f85-n8czk to kind-worker
  Normal   Created           4m19s (x2 over 5m19s)  kubelet            Created container antrea-controller
  Normal   Started           4m19s (x2 over 5m19s)  kubelet            Started container antrea-controller
  Normal   Killing           4m19s                  kubelet            Container antrea-controller failed liveness probe, will be restarted
  Warning  Unhealthy         89s (x8 over 4m59s)    kubelet            Liveness probe failed: Get "https://localhost:10349/livez": dial tcp [::1]:10349: connect: connection refused
  Warning  Unhealthy         79s (x10 over 4m59s)   kubelet            Readiness probe failed: Get "https://localhost:10349/readyz": dial tcp [::1]:10349: connect: connection refused
  Normal   Pulled            19s (x4 over 5m19s)    kubelet            Container image "antrea/antrea-controller-ubuntu-coverage:latest" already present on machine

Reply:
@antoninbas I reproduced the error locally, here is the `kubectl describe` output for that pod:
```
I0515 03:27:30.333084 1309413 framework.go:2660] ERORRRRRR: Name:                 antrea-agent-5dxnn
Namespace:            kube-system
Priority:             2000001000
Priority Class Name:  system-node-critical
Service Account:      antrea-agent
Node:                 kind-worker2/172.18.0.2
Start Time:           Tue, 14 May 2024 21:57:06 +0000
Labels:               app=antrea
                      component=antrea-agent
                      controller-revision-hash=75f74c6b6d
                      pod-template-generation=1
Annotations:          checksum/config: d0b905f11f84008ec883b25acf2106020200fc226e51ca647d0083956e532ae0
                      kubectl.kubernetes.io/default-container: antrea-agent
Status:               Running
IP:                   172.18.0.2
IPs:
  IP:           172.18.0.2
Controlled By:  DaemonSet/antrea-agent
Init Containers:
  install-cni:
    Container ID:  containerd://654efbe0b9208d7127c4b2f1d2fdfdb2fc6b12f78515d84915cf3b133c295cfd
    Image:         antrea/antrea-agent-ubuntu-coverage:latest
    Image ID:      docker.io/library/import-2024-05-14@sha256:ab5b4129a4ee158f3ddeae628c1bf6c161a7b51d429cdcebc3a5b408ff05e2e2
    Port:          <none>
    Host Port:     <none>
    Command:
      install_cni
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 14 May 2024 21:57:06 +0000
      Finished:     Tue, 14 May 2024 21:57:06 +0000
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:  100m
    Environment:
      SKIP_CNI_BINARIES:  
    Mounts:
      /etc/antrea/antrea-cni.conflist from antrea-config (ro,path="antrea-cni.conflist")
      /host/etc/cni/net.d from host-cni-conf (rw)
      /host/opt/cni/bin from host-cni-bin (rw)
      /lib/modules from host-lib-modules (ro)
      /var/run/antrea from host-var-run-antrea (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgqfg (ro)
Containers:
  antrea-agent:
    Container ID:  containerd://e514acc78e315282fde3253a7f24d061006af88c1ac3ff4b60cefcdd46c1c0b6
    Image:         antrea/antrea-agent-ubuntu-coverage:latest
    Image ID:      docker.io/library/import-2024-05-14@sha256:ab5b4129a4ee158f3ddeae628c1bf6c161a7b51d429cdcebc3a5b408ff05e2e2
    Port:          10350/TCP
    Host Port:     10350/TCP
    Args:
      antrea-agent-coverage
      --config=/etc/antrea/antrea-agent.conf
      --logtostderr=false
      --log_dir=/var/log/antrea
      --alsologtostderr
      --log_file_max_size=100
      --log_file_max_num=4
      --v=4
    State:          Running
      Started:      Tue, 14 May 2024 21:57:07 +0000
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:      200m
    Liveness:   http-get https://localhost:api/livez delay=10s timeout=5s period=10s #success=1 #failure=5
    Readiness:  http-get https://localhost:api/readyz delay=5s timeout=5s period=10s #success=1 #failure=8
    Environment:
      POD_NAME:       antrea-agent-5dxnn (v1:metadata.name)
      POD_NAMESPACE:  kube-system (v1:metadata.namespace)
      NODE_NAME:       (v1:spec.nodeName)
    Mounts:
      /etc/antrea/antrea-agent.conf from antrea-config (ro,path="antrea-agent.conf")
      /host/proc from host-proc (ro)
      /host/var/run/netns from host-var-run-netns (ro)
      /run/xtables.lock from xtables-lock (rw)
      /var/lib/cni from host-var-run-antrea (rw,path="cni")
      /var/log/antrea from host-var-log-antrea (rw)
      /var/run/antrea from host-var-run-antrea (rw)
      /var/run/openvswitch from host-var-run-antrea (rw,path="openvswitch")
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgqfg (ro)
  antrea-ovs:
    Container ID:  containerd://999590d3b1960cc24c9412bbe7455c19ba66d6eeeca9972e9e719c3d162954a3
    Image:         antrea/antrea-agent-ubuntu-coverage:latest
    Image ID:      docker.io/library/import-2024-05-14@sha256:ab5b4129a4ee158f3ddeae628c1bf6c161a7b51d429cdcebc3a5b408ff05e2e2
    Port:          <none>
    Host Port:     <none>
    Command:
      start_ovs
    Args:
      --log_file_max_size=100
      --log_file_max_num=4
    State:          Running
      Started:      Tue, 14 May 2024 21:57:07 +0000
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:        200m
    Liveness:     exec [/bin/sh -c timeout 10 container_liveness_probe ovs] delay=5s timeout=10s period=10s #success=1 #failure=5
    Environment:  <none>
    Mounts:
      /var/log/openvswitch from host-var-log-antrea (rw,path="openvswitch")
      /var/run/openvswitch from host-var-run-antrea (rw,path="openvswitch")
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgqfg (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       True 
  ContainersReady             True 
  PodScheduled                True 
Volumes:
  antrea-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      antrea-config
    Optional:  false
  host-cni-conf:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/cni/net.d
    HostPathType:  
  host-cni-bin:
    Type:          HostPath (bare host directory volume)
    Path:          /opt/cni/bin
    HostPathType:  
  host-proc:
    Type:          HostPath (bare host directory volume)
    Path:          /proc
    HostPathType:  
  host-var-run-netns:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/netns
    HostPathType:  
  host-var-run-antrea:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/antrea
    HostPathType:  DirectoryOrCreate
  host-var-log-antrea:
    Type:          HostPath (bare host directory volume)
    Path:          /var/log/antrea
    HostPathType:  DirectoryOrCreate
  host-lib-modules:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/modules
    HostPathType:  
  xtables-lock:
    Type:          HostPath (bare host directory volume)
    Path:          /run/xtables.lock
    HostPathType:  FileOrCreate
  kube-api-access-qgqfg:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 :NoSchedule op=Exists
                             :NoExecute op=Exists
                             CriticalAddonsOnly op=Exists
                             node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/network-unavailable:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists
                             node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                             node.kubernetes.io/unreachable:NoExecute op=Exists
                             node.kubernetes.io/unschedulable:NoSchedule op=Exists
Events:
  Type     Reason     Age   From               Message
  ----     ------     ----  ----               -------
  Normal   Scheduled  24s   default-scheduler  Successfully assigned kube-system/antrea-agent-5dxnn to kind-worker2
  Normal   Pulled     24s   kubelet            Container image "antrea/antrea-agent-ubuntu-coverage:latest" already present on machine
  Normal   Created    24s   kubelet            Created container install-cni
  Normal   Started    24s   kubelet            Started container install-cni
  Normal   Pulled     23s   kubelet            Container image "antrea/antrea-agent-ubuntu-coverage:latest" already present on machine
  Normal   Created    23s   kubelet            Created container antrea-agent
  Normal   Started    23s   kubelet            Started container antrea-agent
  Normal   Pulled     23s   kubelet            Container image "antrea/antrea-agent-ubuntu-coverage:latest" already present on machine
  Normal   Created    23s   kubelet            Created container antrea-ovs
  Normal   Started    23s   kubelet            Started container antrea-ovs
  Warning  Unhealthy  14s   kubelet            Readiness probe failed: Get "https://localhost:10350/readyz": dial tcp [::1]:10350: connect: connection refused
```